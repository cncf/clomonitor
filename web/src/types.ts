export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
