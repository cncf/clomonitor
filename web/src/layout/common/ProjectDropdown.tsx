import classNames from 'classnames';
import { MouseEvent as ReactMouseEvent, useRef, useState } from 'react';
import { GoThreeBars } from 'react-icons/go';

import useOutsideClick from '../../hooks/useOutsideClick';
import { Foundation } from '../../types';
import BadgeModal from './BadgeModal';
import styles from './ProjectDropdown.module.css';
import ReportSummaryModal from './ReportSummaryModal';

interface OpenModalStatus {
  status: boolean;
  name?: string;
}

enum Modals {
  Badge = 'badge',
  ReportSummary = 'reportSummary',
}

interface Props {
  foundation: Foundation;
  orgName: string;
  projectName: string;
  projectDisplayName?: string;
}

const ProjectDropdown = (props: Props) => {
  const ref = useRef(null);
  const [visibleDropdown, setVisibleDropdown] = useState<boolean>(false);
  const [openStatus, setOpenStatus] = useState<OpenModalStatus>({ status: false });
  useOutsideClick([ref], visibleDropdown, () => setVisibleDropdown(false));

  const onCloseModal = () => {
    setOpenStatus({ status: false });
  };

  return (
    <>
      <div ref={ref} className="ms-auto position-relative">
        <button
          data-testid="dropdown-btn"
          type="button"
          className={`btn btn-sm btn-primary text-white rounded-0 lh-1 ${styles.btn}`}
          onClick={(e: ReactMouseEvent<HTMLButtonElement, MouseEvent>) => {
            e.preventDefault();
            e.stopPropagation();
            setVisibleDropdown(!visibleDropdown);
          }}
        >
          <GoThreeBars />
        </button>

        <ul
          role="complementary"
          className={classNames('dropdown-menu rounded-0', styles.dropdown, { show: visibleDropdown })}
        >
          <li>
            <button
              className="dropdown-item"
              onClick={(e) => {
                e.stopPropagation();
                e.preventDefault();

                setVisibleDropdown(false);
                setOpenStatus({ name: Modals.Badge, status: true });
              }}
            >
              Get badge
            </button>
          </li>
          <li>
            <button
              className="dropdown-item"
              onClick={(e) => {
                e.stopPropagation();
                e.preventDefault();

                setVisibleDropdown(false);
                setOpenStatus({ name: Modals.ReportSummary, status: true });
              }}
            >
              Embed report summary
            </button>
          </li>
        </ul>
      </div>

      <BadgeModal
        foundation={props.foundation}
        orgName={props.orgName}
        projectName={props.projectName}
        openStatus={openStatus}
        onCloseModal={onCloseModal}
      />

      <ReportSummaryModal
        foundation={props.foundation}
        orgName={props.orgName}
        projectName={props.projectName}
        openStatus={openStatus}
        onCloseModal={onCloseModal}
      />
    </>
  );
};

export default ProjectDropdown;
