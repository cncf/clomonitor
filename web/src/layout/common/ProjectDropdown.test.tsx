import { fireEvent, render, screen } from '@testing-library/react';

import { Foundation } from '../../types';
import ProjectDropdown from './ProjectDropdown';

const defaultProps = {
  foundation: Foundation.cncf,
  orgName: 'org',
  projectName: 'proj',
};

describe('ProjectDropdown', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<ProjectDropdown {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders proper content', () => {
    render(<ProjectDropdown {...defaultProps} />);

    const dropdown = screen.getByRole('complementary');
    expect(dropdown).not.toHaveClass('show');

    const btn = screen.getByTestId('dropdown-btn');
    expect(btn).toBeInTheDocument();

    fireEvent.click(btn);

    expect(dropdown).toHaveClass('show');
    expect(screen.getByRole('button', { name: 'Get badge' })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: 'Embed report summary' })).toBeInTheDocument();
  });

  it('opens Get badge modal', () => {
    render(<ProjectDropdown {...defaultProps} />);

    const dropdown = screen.getByRole('complementary');
    expect(dropdown).not.toHaveClass('show');

    fireEvent.click(screen.getByTestId('dropdown-btn'));

    expect(dropdown).toHaveClass('show');
    const badgeBtn = screen.getByRole('button', { name: 'Get badge' });
    fireEvent.click(badgeBtn);

    expect(dropdown).not.toHaveClass('show');
    expect(screen.getByText('Project badge')).toBeInTheDocument();
  });

  it('opens Report summary modal', () => {
    render(<ProjectDropdown {...defaultProps} />);

    const dropdown = screen.getByRole('complementary');
    expect(dropdown).not.toHaveClass('show');

    fireEvent.click(screen.getByTestId('dropdown-btn'));

    expect(dropdown).toHaveClass('show');
    const reportBtn = screen.getByRole('button', { name: 'Embed report summary' });
    fireEvent.click(reportBtn);

    expect(dropdown).not.toHaveClass('show');
    expect(screen.getAllByText('Embed report summary')).toHaveLength(2);
  });
});
