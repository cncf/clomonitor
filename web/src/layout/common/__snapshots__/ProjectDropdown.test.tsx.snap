// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectDropdown creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="ms-auto position-relative"
  >
    <button
      class="btn btn-sm btn-primary text-white rounded-0 lh-1 btn"
      data-testid="dropdown-btn"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M14 5H2V3h12v2zm0 4H2V7h12v2zM2 13h12v-2H2v2z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <ul
      class="dropdown-menu rounded-0 dropdown"
      role="complementary"
    >
      <li>
        <button
          class="dropdown-item"
        >
          Get badge
        </button>
      </li>
      <li>
        <button
          class="dropdown-item"
        >
          Embed report summary
        </button>
      </li>
    </ul>
  </div>
  <div>
    <div
      aria-modal="false"
      class="modal Modal-module_modal__aD7aP"
      role="dialog"
    >
      <div
        class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
      >
        <div
          class="modal-content rounded-0 border border-3 mx-auto Modal-module_content__dWO-B"
        >
          <div
            class="modal-header rounded-0 d-flex flex-row align-items-center Modal-module_header__ILG9i"
          >
            <div
              class="modal-title h5 m-2 flex-grow-1"
            >
              Project badge
            </div>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body p-4 h-100 d-flex flex-column"
          />
          <div
            class="modal-footer p-3"
          >
            <button
              aria-label="Close modal"
              class="btn btn-sm rounded-0 btn-secondary text-uppercase"
              type="button"
            >
              <div
                class="d-flex flex-row align-items-center"
              >
                <svg
                  class="me-2"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
                <div>
                  Close
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      aria-modal="false"
      class="modal Modal-module_modal__aD7aP"
      role="dialog"
    >
      <div
        class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
      >
        <div
          class="modal-content rounded-0 border border-3 mx-auto Modal-module_content__dWO-B"
        >
          <div
            class="modal-header rounded-0 d-flex flex-row align-items-center Modal-module_header__ILG9i"
          >
            <div
              class="modal-title h5 m-2 flex-grow-1"
            >
              Embed report summary
            </div>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body p-4 h-100 d-flex flex-column"
          />
          <div
            class="modal-footer p-3"
          >
            <button
              aria-label="Close modal"
              class="btn btn-sm rounded-0 btn-secondary text-uppercase"
              type="button"
            >
              <div
                class="d-flex flex-row align-items-center"
              >
                <svg
                  class="me-2"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
                <div>
                  Close
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
